File: scripts/ci/check-infra-manifest-registry.py
#!/usr/bin/env python3
import re
import subprocess
import sys
from pathlib import Path

README = Path('infra/README.md')


def parse_registry(text: str):
    section_match = re.search(r"## Active vs Deprecated Infra Manifest Registry\n(.*?)(?:\n## |\Z)", text, re.S)
    if not section_match:
        raise ValueError('Missing "Active vs Deprecated Infra Manifest Registry" section in infra/README.md')

    rows = []
    for line in section_match.group(1).splitlines():
        if not line.strip().startswith('|'):
            continue
        cols = [c.strip() for c in line.strip().strip('|').split('|')]
        if len(cols) < 4 or cols[0] in {'Type', '---'}:
            continue
        rows.append(cols)
    return rows


def main():
    if not README.exists():
        print('[infra-registry] infra/README.md not found')
        return 1

    text = README.read_text(encoding='utf-8')
    try:
        rows = parse_registry(text)
    except ValueError as e:
        print(f'[infra-registry] {e}')
        return 1

    if not rows:
        print('[infra-registry] Registry is empty')
        return 1

    deprecated = []
    missing = []
    for row in rows:
        _, path, lifecycle, *_ = row + [''] * (4 - len(row))
        manifest_path = path.strip('`')
        lifecycle = lifecycle.lower()
        if lifecycle == 'deprecated':
            deprecated.append(manifest_path)
        if lifecycle == 'active' and not Path(manifest_path).exists():
            missing.append(manifest_path)

    if missing:
        print('[infra-registry] Active entries that do not exist:')
        for m in missing:
            print(f'  - {m}')
        return 1

    offenders = []
    for dep in deprecated:
        cmd = [
            'rg', '--line-number', '--fixed-strings', dep,
            '--glob', '!.git/**',
            '--glob', '!infra/README.md',
            '--glob', '!scripts/ci/check-infra-manifest-registry.py',
            '.'
        ]
        result = subprocess.run(cmd, text=True, capture_output=True)
        if result.returncode == 0 and result.stdout.strip():
            offenders.append((dep, result.stdout.strip().splitlines()))

    if offenders:
        print('[infra-registry] Deprecated manifest paths are still referenced:')
        for dep, lines in offenders:
            print(f'  - {dep}')
            for line in lines[:5]:
                print(f'    {line}')
        return 1

    print('[infra-registry] Registry is valid and no deprecated paths are referenced.')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/ci/check-infra-manifest-registry.mjs
#!/usr/bin/env node
import { execSync } from 'node:child_process';
import { readFileSync } from 'node:fs';

const REGISTRY_FILE = 'infra/README.md';
const START = '<!-- infra-manifest-registry:start -->';
const END = '<!-- infra-manifest-registry:end -->';

function run(cmd) {
  return execSync(cmd, { encoding: 'utf8', stdio: ['ignore', 'pipe', 'pipe'] }).trim();
}

function parseRegistry(markdown) {
  const start = markdown.indexOf(START);
  const end = markdown.indexOf(END);
  if (start === -1 || end === -1 || end <= start) {
    throw new Error(`Registry markers not found in ${REGISTRY_FILE}.`);
  }

  const block = markdown.slice(start + START.length, end);
  const lines = block
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.startsWith('|') && !line.includes('---'));

  const rows = [];
  for (const line of lines) {
    const cols = line.split('|').map((c) => c.trim()).filter(Boolean);
    if (cols.length < 4 || cols[0] === 'Path') continue;
    rows.push({ path: cols[0].replace(/`/g, ''), status: cols[1].toLowerCase(), replacement: cols[2], notes: cols[3] });
  }
  return rows;
}

function findDeprecatedReferences(deprecatedPath) {
  try {
    const result = run([
      'rg --line-number --fixed-strings',
      `"${deprecatedPath}"`,
      '--glob "!infra/README.md"',
      '--glob "!scripts/ci/check-infra-manifest-registry.mjs"',
      '--glob "!.git/**"',
      '.',
    ].join(' '));
    return result ? result.split('\n') : [];
  } catch (error) {
    if (error.status === 1) return [];
    throw error;
  }
}

try {
  const content = readFileSync(REGISTRY_FILE, 'utf8');
  const registry = parseRegistry(content);

  if (registry.length === 0) {
    throw new Error('Registry is empty. Add at least one manifest row.');
  }

  const activeCount = registry.filter((r) => r.status === 'active').length;
  const deprecated = registry.filter((r) => r.status === 'deprecated');

  if (activeCount === 0) {
    throw new Error('Registry must contain at least one active manifest entry.');
  }

  const failures = [];
  for (const entry of deprecated) {
    const refs = findDeprecatedReferences(entry.path);
    if (refs.length > 0) {
      failures.push(`Deprecated path ${entry.path} is still referenced:\n${refs.map((r) => `  - ${r}`).join('\n')}`);
    }
  }

  if (failures.length > 0) {
    console.error('❌ Infra manifest registry check failed.');
    for (const failure of failures) console.error(`\n${failure}`);
    process.exit(1);
  }

  console.log(`✅ Infra manifest registry check passed (${activeCount} active, ${deprecated.length} deprecated entries).`);
} catch (error) {
  console.error(`❌ Infra manifest registry check failed: ${error.message}`);
  process.exit(1);
}
